### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: GcWatchTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+public class GcWatchTest
+{
+    private static final List<byte[]> VALUES = new ArrayList<>();
+    private static final int VALUE_SIZE = 100 * 1024;
+    private static final long MAX_MEMORY_SIZE = (long) (0.9 * 1024 * 1024 * 1024);
```
Comment (dain) [inline GcWatchTest.java:23]: Include the required command line arguments

File: README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Overview
+
+**gcwatch** is a simple [JVM TI][] agent that monitors stop-the-world
+garbage collections and makes them available via a simple TCP server
+interface. This allows detecting if the machine and Java process is still
```
Comment (dain) [inline README.md:5]: `allows easy detection`?

File: gcwatch.c [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+static void JNICALL gcStart(jvmtiEnv *jvmti)
+{
+   start = getMillis();
+   inGc = true;
+}
+
+static void JNICALL gcFinish(jvmtiEnv *jvmti)
+{
+   inGc = false;
```
Comment (dain) [inline gcwatch.c]: maybe reset start to zero, so it is symmetric  with the `gcStart` method?  Doesn't really matter.

File: gcwatch.c [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+   }
+
+   struct sockaddr_in6 addr = {
+      .sin6_family = AF_INET6,
+      .sin6_port = htons(port),
+      .sin6_addr = in6addr_any,
+   };
+
+   if (bind(server, (struct sockaddr *) &addr, sizeof(addr)) == -1) {
+      perror("ERROR: failed to bind GcWatch socket");
```
Comment (dain) [inline gcwatch.c]: can you add the port to the message?

File: gcwatch.c [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+   if (listen(server, SOMAXCONN) == -1) {
+      perror("ERROR: failed to listen on GcWatch socket");
+      exit(1);
+   }
+
+   fprintf(stderr, "GcWatch listener started on port %d\n", port);
+   
+   while (true) {
+      int client = accept(server, NULL, 0);
+      if (client == -1) {
```
Comment (dain) [inline gcwatch.c]: the man page says:

       On success, these system calls return a nonnegative integer that is a
       file descriptor for the accepted socket.  On error, -1 is returned,
       and errno is set appropriately.

Should this check `< 0`?

File: gcwatch.c [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+      write(client, output, strlen(output));
+      close(client);
+   }
+}
+
+static void JNICALL vmInit(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread)
+{
+   jclass clazz = (*jni)->FindClass(jni, "java/lang/Thread");
+   if (clazz == NULL) {
+      fprintf(stderr, "ERROR: failed to find Thread class\n");
```
Comment (dain) [inline gcwatch.c]: maybe add `gcWatch` to every message here

File: gcwatch.c [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+      }
+      
+      write(client, output, strlen(output));
+      close(client);
+   }
+}
+
+static void JNICALL vmInit(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread)
+{
+   jclass clazz = (*jni)->FindClass(jni, "java/lang/Thread");
```
Comment (dain) [inline gcwatch.c]: Maybe put this thread construction in a separate method, or add a block comment explaining what you are doing.

File: gcwatch.c [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+      return JNI_ERR;
+   }
+
+   jvmtiCapabilities capabilities = {
+      .can_generate_garbage_collection_events = true,
+   };
+
+   err = (*jvmti)->AddCapabilities(jvmti, &capabilities);
+   if (err != JVMTI_ERROR_NONE) {
+      fprintf(stderr, "ERROR: AddCapabilities failed: %d\n", err);
```
Comment (dain) [inline gcwatch.c]: Maybe mention "cannot enable GC events" or such?

File: gcwatch.c [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+      return JNI_ERR;
+   }
+
+   jvmtiEvent events[] = {
+      JVMTI_EVENT_VM_INIT,
+      JVMTI_EVENT_GARBAGE_COLLECTION_START,
+      JVMTI_EVENT_GARBAGE_COLLECTION_FINISH,
+   };
+
+   for (int i = 0; i < 3; i++) {
```
Comment (dain) [inline gcwatch.c]: hard coded `3`

**Review by dain — APPROVED** [review]
Very Nice!

File: gcwatch.c [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <jvmti.h>
+
+static int port;
+
+static bool inGc;
+static long start;
```
Comment (dain) [inline gcwatch.c]: Are `bool` and `long` atomic in C?

File: gcwatch.c (hunk 3 above)
Comment (dain) [reply gcwatch.c]: I guess that there could be a race condition between the call back  if we reset start to 0

File: gcwatch.c [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+{
+   struct timespec ts;
+   clock_gettime(CLOCK_MONOTONIC, &ts);
+   return (ts.tv_sec * 1000) + (ts.tv_nsec / (1000 * 1000));
+}
+
+static void JNICALL gcStart(jvmtiEnv *jvmti)
+{
+   start = getMillis();
+   inGc = true;
```
Comment (dain) [inline gcwatch.c]: maybe do the `inGc = true;` to reduce the impact of races

File: gcwatch.c (hunk 4 above)
Comment (electrum) [reply gcwatch.c]: That's a good idea, but `perror` doesn't support that, and `strerror` isn't thread safe, so I'd rather keep the code simple.

File: gcwatch.c (hunk 5 above)
Comment (electrum) [reply gcwatch.c]: Checking for `-1` is canonical for UNIX system calls. The "nonnegative" is simply explaining the requirement for returned file descriptors.

File: gcwatch.cpp [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+      exit(1);
+   }
+
+   jstring name = jni->NewStringUTF(threadName);
+   if (name == NULL) {
+      fprintf(stderr, "ERROR: GcWatch: failed to create String object\n");
+      exit(1);
+   }
+
+   jthread object = jni->NewObject(clazz, init, name);
```
Comment (dain) [inline gcwatch.cpp:130]: do we need to set daemon for this thread?

**Review by dain — APPROVED** [review]
looks good

File: gcwatch.cpp (hunk 12 above)
Comment (electrum) [reply gcwatch.cpp:130]: `RunAgentThread` creates daemon threads

**Pull request merged**